충남대 BE_이은경 4주차 과제 (2단계)

pkyung opened the pull request:
## 과제 진행 사항
### 상품 옵션
- [x] 옵션 도메인 설계
- [x] 옵션 dto validation 설정
- [x] 상품과 옵션 설정
- [x] 옵션 cr 구현
- [x] 옵션 이름 중복 확인 메서드 생성 (도메인 계층에 생성)

pkyung committed:
Chore: 코드 옮기기 - step0

pkyung committed:
Docs: 기능 목록 작성 - step1

pkyung committed:
Feat: 카테고리 도메인 설계

카테고리 엔티티 설계를 구현했다. category의 필드는 id, name, color,
imageUrl, description으로 이뤄져 있으며 모든 필드는 null이 불가하도록
설정했다.

pkyung committed:
Feat: 카테고리의 관계 설정

카테고리는 상품과 1대1 관계라고 생각했다. 따라서, Product 도메인에
@OneToOne 관계로 설정을 해주었다. Category에는 아직 상품을 확인할 필요가
없을 것 같아서 양방향 매핑을 하지 않았다.

pkyung committed:
Feat: 카테고리 ru 구현

카테고리는 상품과 함께 지정할 수 있다는 요구사항이 있다. 따라서 Product
생성자와 update 를 category를 넣은 채로 수정했다.
그리고 category는 수정이 가능하고 모든 category를 조회하는 api가
요구사항에 존재한다. 따라서, findAllCategories와 updateCategory 메서드를
만들어서 카테고리를 조회하고 수정하도록 했다.

pkyung committed:
Refactor: handler 클래스 정리

notfound, alreadyExists, MethodArgument, Unauthorized exception 들을
다루는 handler로 클래스를 정리했다.

pkyung committed:
Chore: 안 옮긴 test 코드 옮김

pkyung committed:
Fix: 상품에 카테고리는 필수로 들어가야 해서 수정함

pkyung committed:
Fix: @ManyToOne으로 수정함

pkyung committed:
Feat: 화면 반영 - 상품과 함께 카테고리  추가되도록

Product를 Page<Product>로 반환하면 Product만 반환되고 연관관계 설정된
Category는 나오지 않아서 ProductResponse를 생성하여 반환하도록 수정헸다.
Product와 Category에 필요한 getter들을 생성했다.

pkyung committed:
Refactor: 필요없는 import 삭제

pkyung committed:
Refactor: 하나만 조회하는 api 또한 productResponse 반환

pkyung committed:
Refactor: token 관련 메서드 import 해서 사용

pkyung committed:
Fix: wish.js의 createWish 전역 변수 설정

pkyung committed:
Refactor: category getter 추가

pkyung committed:
Test: 상품과 카테고리 저장 테스트

pkyung committed:
Docs: 기능 목록 구현 - step1

pkyung committed:
Merge branch 'pkyung' into main

pkyung committed:
Refactor: merge 시 삭제 안된 파일 삭제

pkyung committed:
Refactor: 불필요한 import 제거

pkyung committed:
Refactor: merge 시 삭제 안된 파일 삭제

pkyung committed:
Refactor: category controller와 service 리펙토링

1. requestMapping에 /api/categories 로 수정했다.
2. 메서드명 updateCategory -> update
3. category list의 변수병을 복수형으로 수정했다.
4. update 의 경우 반환할 때 no_content
5. body가 없는 responseEntity는 Void 제네릭 타입으로 명시했다.

pkyung committed:
Refactor: request validation 추가

1. ProductRequest와 CategoryRequest의 color 필드를 정규식으로 검증했다.
2. imageUrl 필드를 정규식을 통해 검증했다.

pkyung committed:
Refactor: productResponse 계층형으로 수정

1. categoryName, categoryId를 category 객체로 수정했다.
2. service의 반환값의 생성자 일부를 수정했다.

pkyung committed:
Refactor: category name 유니크 추가

pkyung committed:
Refactor: orElseThrow로 optional 뗌

pkyung committed:
Refactor: category 엔티티 validate 설정 추가

pkyung committed:
Refactor: description 필드 255자 이상을 위해 @Lob 추가

pkyung committed:
Refactor: Exception 관리

1. AlreadyExistsException과 NotFoundException 을 생성했다.
2. Service에 사용된 Exception들을 수정했다.
3. GlobalExceptionHandler에서 exception을 관리하도록 수정했다.

pkyung committed:
Refactor: 메서드 이름 수정

pkyung committed:
Refactor: Optional orElseThrow로 뗌

pkyung committed:
Docs: 기능 목록 작성 - step2

pkyung committed:
Feat: 옵션 도메인 설계

option의 validation과 필드를 설계했다.

pkyung committed:
Refactor: length가 있기 때문에 notnull이 의미없음

pkyung committed:
Feat: 옵션 dto validation 설정

옵션의 name 필드는 특정한 특수문자만 들어갈 수 있어서 정규식을 통해
설정했다. 그리고 quatity 필드는 최소 1 최대 1억 미만이므로 min max
설정을 했다.

pkyung committed:
Feat: 상품과 옵션 관계 설정

옵션과 상품은 ManyToOne의 관계이다. 따라서, 옵션에는 ManyToOne 매핑을,
상품에는 OneToMany 옵션을 추가했다. 그리고 옵션에 product와의 관계
설정을 위한 setProduct 메서드를 추가하여 product의 options 리스트에
객체를 추가하며 관계를 설정하도록 구현했다.

pkyung committed:
Fix: responseBody와 valid 어노테이션 추가

pkyung committed:
Feat: 옵션 이름 중복 확인 메서드 생성

1. 옵션에 getName 메서드 추가했다.
2. product에 options 필드의 name과 동일한 이름이 있는지 확인하는
   checkDuplicateOptionName 메서드를 추가했다.

pkyung committed:
Feat: 옵션 cr 구현

옵션을 생성하고 조회하도록 구현했다.
생성의 경우, productId와 함께 입력 받아 옵션 생성과 관계 설정을 함께
하도록 구현했다.

pkyung committed:
Fix: pathVariable 인식 안되어 문자열 추가

pkyung committed:
Refactor: imageUrl 정규식 수정함

pkyung committed:
Fix: pathvariable 매개변수 추가

pkyung committed:
Fix: 데이터 직렬화를 위해 getter 추가

pkyung committed:
Refactor: 필요 없는 import 제거

pkyung committed:
Merge branch 'pkyung' into main

pkyung committed:
Docs: 기능 목록 구현 - step2

pkyung committed:
Test: 옵션 테스트 추가

ksy90101 commented on src/main/java/gift/dto/OptionRequest.java:
```diff
@@ -0,0 +1,17 @@
+        @Length(max = 50, message = "옵션 이름에는 공백을 포함하여 최대 50자까지 입력할 수 있습니다.")
+        String name,
+        @Min(1)
+        @Max(99999999)
```
max는 지정 안해도 괜찮을거 같기도 해요~
굳이 정해지지 않았다면요!

pkyung replied:
1억개 미만이라는 요구사항이 있어서 넣었습니다!

ksy90101 commented on src/main/java/gift/model/Product.java:
```diff
@@ -71,4 +76,15 @@ public String getImageUrl() {
+    }
+    public boolean checkDuplicateOptionName(String optionName) {
+        for (Option option : options) {
+            if (option.getName().equals(optionName)) {
```
해당 코드를 option 메서드로 구현해서 처리하면 어떨까요?

pkyung replied:
네 optoin으로 수정하겠습니다. product에서 한 이유는 option이 아직 생성되기 이전이었기 때문이었는데 option의  생성자를 통해 product를 정의하면 option에서 메서드로 구현할 수 있을 것 같습니다.

pkyung replied:
생성자에서 처리하니 option 객체 생성 시에 이미 product의 options에 option이 들어가서 동일한 이름으로 체킹 되어서 product에서 해야할 것 같습니다.

ksy90101 replied:
차라리 지금 보니 SQL 쿼리로 처리해도 무관할거 같기도 하네요.

ksy90101 replied:
혹은 product 도메인 로직으로 있는게 어울릴거 같아요,.
product의 options의 validate 느낌이죠.

ksy90101 commented on src/main/java/gift/model/Product.java:
```diff
@@ -71,4 +76,15 @@ public String getImageUrl() {
+        return options;
+    }
+    public boolean checkDuplicateOptionName(String optionName) {
+        for (Option option : options) {
```
stream을 사용해볼수도 있을거 같은데 한번 해보시면 어떨까요?

pkyung replied:
네 stream을 통해 처리해보겠습니다.

ksy90101 commented on src/main/java/gift/service/OptionService.java:
```diff
@@ -0,0 +1,45 @@
+    public Option makeOption(Long productId, OptionRequest request) {
+        Product product = productRepository.findById(productId)
+                .orElseThrow(() -> new NotFoundException("해당 id의 상품이 존재하지 않습니다."));
+        if (product.checkDuplicateOptionName(request.name())) {
+            throw new AlreadyExistsException("해당 상품에 등록된 옵션과 동일한 이름이 존재합니다.");
```
👍

ksy90101 commented on src/main/java/gift/service/OptionService.java:
```diff
@@ -0,0 +1,45 @@
+        if (product.checkDuplicateOptionName(request.name())) {
+            throw new AlreadyExistsException("해당 상품에 등록된 옵션과 동일한 이름이 존재합니다.");
+        }
+        Option option = new Option(
+                request.name(),
+                request.quantity()
+        );
+        option.setProduct(product);
```
생성자에서 처리해볼수도 있을거 같아요?

ksy90101 commented on src/test/java/gift/JpaOptionTest.java:
```diff
@@ -0,0 +1,64 @@
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+@DataJpaTest
+public class JpaOptionTest {
```
JPA 테스트는 이제 잘 되는거 같아요~
한번 이번엔 Service Test도 추가해보면 어떨까여?

pkyung replied:
네 추가해보겠습니다.

ksy90101 reviewed (changes requested):
깔끔하게 잘 구현해주셨네요 👍 
고생많으셨습니다!
간단한 피드백을 남겼으니 확인해주세요~
질문이 있다면 언제든 DM주세요!

pkyung committed:
Refactor: 멘토님 피드백 추가

checkDuplicateOptionName 메서드를 option에 추가했다. 그리고 stream을
통해 더 간결하게 작성했습니다.
option과 product 사이의 관계를 생성자를 통해 설정했다.

pkyung committed:
Fix: checkDuplicate 위치 재변경

pkyung committed:
Fix: 재변경

pkyung committed:
Test: option service 테스트 구현

option service의 getOptionByProductId와 makeOption을 테스트했다.

pkyung committed:
Refactor: stream으로 수정함

ksy90101 commented on src/test/java/gift/OptionServiceTest.java:
```diff
@@ -0,0 +1,100 @@
+package gift;
```
테스트 파일도 실제 프로덕션의 패키지 구조와 동일하게 가져갑니다.
지금 service가 service폴더 내에 있으니 테스트도 동일하게 패키지 구조를 가져가면 좋겠습니다.

pkyung replied:
네 수정하겠습니다!

ksy90101 commented on src/test/java/gift/OptionServiceTest.java:
```diff
@@ -0,0 +1,100 @@
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+public class OptionServiceTest {
```
👍

ksy90101 reviewed (approved):
2단계도 고생많으셨어요~
간단한 피드백을 남겼으니 확인해주세요!
추가적으로 궁금한게 있다면 DM주세요!
